File: src/cache.ts
import { type Span, SpanStatusCode, trace } from "@opentelemetry/api";
import { Client } from "hazelcast-client";

/**
 * The app connects as a Hazelcast *client* to the standalone Hazelcast member deployed
 * on its own node (see k8s/hazelcast-deployment.yaml), rather than embedding a member in
 * each pod - keeps the cache tier independent of app pod restarts/scaling. Unisocket mode
 * (smartRouting: false) matches the old Java client config: there's a single member behind
 * a plain ClusterIP Service, so smart routing (talking to every member directly) isn't
 * useful here.
 *
 * Deliberately no Near Cache, for the same reason documented in the old
 * HazelcastConfig.java: a client-side Near Cache on this map went stale across pods after
 * an evict() - the invalidation broadcast wasn't reliably reaching every other client's
 * near-cache. Reads go straight to the shared Hazelcast member instead, which stays correct.
 *
 * Values are cached as JSON strings (GraphQL response shape - id/title/content/author/
 * createdAt/version - see serializeMessage in resolvers.ts), keyed by message id.
 */
const CACHE_NAME = "message-service-cache";
const MAP_NAME = "messages";

let client: Client | undefined;

// Uses the global tracer provider registered in src/tracing.ts - a no-op until that runs.
const tracer = trace.getTracer("message-service");

// Hazelcast has no OpenTelemetry instrumentation, so each cache call gets a manual span
// (a child of the GraphQL resolver span). Without them a cache hit shows up in a trace as a
// resolver with no Prisma spans, and the network hop to the Hazelcast member is an
// unexplained gap inside the resolver's duration.
async function traced<T>(operation: string, fn: (span: Span) => Promise<T>): Promise<T> {
  return tracer.startActiveSpan(
    `hazelcast.${operation}`,
    { attributes: { "db.system": "hazelcast", "db.operation": operation, "cache.map": MAP_NAME } },
    async (span) => {
      try {
        return await fn(span);
      } catch (err) {
        span.recordException(err as Error);
        span.setStatus({ code: SpanStatusCode.ERROR, message: (err as Error).message });
        throw err;
      } finally {
        span.end();
      }
    },
  );
}

export async function connectCache(host: string, port: string): Promise<void> {
  client = await Client.newHazelcastClient({
    clusterName: CACHE_NAME,
    network: {
      clusterMembers: [`${host}:${port}`],
      smartRouting: false,
    },
  });
}

export function isCacheConnected(): boolean {
  return client !== undefined;
}

export async function getCachedMessage<T>(id: string): Promise<T | null> {
  if (!client) return null;
  const hz = client;
  return traced("get", async (span) => {
    const map = await hz.getMap<string, string>(MAP_NAME);
    const raw = await map.get(id);
    span.setAttribute("cache.hit", Boolean(raw));
    return raw ? (JSON.parse(raw) as T) : null;
  });
}

export async function setCachedMessage(id: string, value: unknown): Promise<void> {
  if (!client) return;
  const hz = client;
  await traced("set", async () => {
    const map = await hz.getMap<string, string>(MAP_NAME);
    await map.set(id, JSON.stringify(value));
  });
}

export async function evictCachedMessage(id: string): Promise<void> {
  if (!client) return;
  const hz = client;
  await traced("delete", async () => {
    const map = await hz.getMap<string, string>(MAP_NAME);
    await map.delete(id);
  });
}

export async function shutdownCache(): Promise<void> {
  await client?.shutdown();
  client = undefined;
}

File: app/cache.py
"""Hazelcast client wrapper (map `messages`, key: message id, value: JSON string of the message).

The app connects as a Hazelcast *client* to the standalone member deployed on its own node
(k8s/hazelcast-deployment.yaml), so the cache tier is independent of app pod restarts/scaling.
Unisocket mode (smart_routing=False) because there is a single member behind a ClusterIP Service.

Deliberately no Near Cache: a client-side Near Cache went stale across pods after an evict(),
because the invalidation broadcast didn't reliably reach every other client. Reads go straight to
the shared member instead. No TTL: entries are evicted after a successful update/delete.

The Python client is thread-based and blocking, so every call runs in a worker thread
(asyncio.to_thread) to keep the event loop free.
"""

import asyncio
from collections.abc import Awaitable, Callable
from typing import Any, Protocol

from hazelcast.client import HazelcastClient
from opentelemetry import trace
from opentelemetry.trace import Span, StatusCode

CLUSTER_NAME = "message-service-cache"
MAP_NAME = "messages"

_tracer = trace.get_tracer("message-service")


class MessageCache(Protocol):
    async def get(self, message_id: str) -> str | None: ...
    async def set(self, message_id: str, value: str) -> None: ...
    async def evict(self, message_id: str) -> None: ...


class HazelcastCache:
    def __init__(self) -> None:
        self._client: HazelcastClient | None = None
        self._map: Any = None

    async def connect(self, host: str, port: int) -> None:
        # The cache is mandatory: if the member is unreachable, startup fails rather than
        # silently running without it. cluster_connect_timeout is bounded (the client default
        # retries for 120s) so a dead member fails well inside the ~70s startup-probe budget.
        def _connect() -> HazelcastClient:
            return HazelcastClient(
                cluster_name=CLUSTER_NAME,
                cluster_members=[f"{host}:{port}"],
                smart_routing=False,
                cluster_connect_timeout=20,
            )

        self._client = await asyncio.to_thread(_connect)
        self._map = self._client.get_map(MAP_NAME).blocking()

    @property
    def connected(self) -> bool:
        return self._client is not None and self._client.lifecycle_service.is_running()

    async def _traced[T](self, operation: str, fn: Callable[[Span], Awaitable[T]]) -> T:
        # Hazelcast has no OpenTelemetry instrumentation, so each cache call gets a manual span
        # (a child of the request span). Otherwise a cache hit shows up as a request with no DB
        # spans and an unexplained gap for the network hop to the member.
        with _tracer.start_as_current_span(
            f"hazelcast.{operation}",
            attributes={
                "db.system": "hazelcast",
                "db.operation": operation,
                "cache.map": MAP_NAME,
            },
        ) as span:
            try:
                return await fn(span)
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    async def get(self, message_id: str) -> str | None:
        async def run(span: Span) -> str | None:
            raw: str | None = await asyncio.to_thread(self._map.get, message_id)
            span.set_attribute("cache.hit", raw is not None)
            return raw

        return await self._traced("get", run)

    async def set(self, message_id: str, value: str) -> None:
        async def run(_: Span) -> None:
            await asyncio.to_thread(self._map.set, message_id, value)

        await self._traced("set", run)

    async def evict(self, message_id: str) -> None:
        async def run(_: Span) -> None:
            await asyncio.to_thread(self._map.delete, message_id)

        await self._traced("delete", run)

    async def shutdown(self) -> None:
        if self._client is not None:
            await asyncio.to_thread(self._client.shutdown)
            self._client = None
            self._map = None


cache = HazelcastCache()


def get_cache() -> MessageCache:
    return cache
